Use existing arch/os information for searches and index operations

estesp opened the pull request:
This code is a first step in making the current registry aware of
images for non-Intel architecture.  While os==linux today, if
we are adding architecture we might as well add both `arch` and
`os` for future use by Microsoft and/or others where os != linux.

Signed-off-by: Pradipta Kr. Banerjee bpradip@in.ibm.com
Signed-off-by: Phil Estes estesp@linux.vnet.ibm.com

estesp commented:
@wking Thought I would get to it before my holiday time off, but didn't--however, I've added a docs PR for the search API change. I also fixed all `flake8` errors from the CI run that didn't show locally when I ran it, but still have CI failures that seem to be unrelated to my code..any input, suggestions welcome!

dmp42 commented:
You can ignore circleci, but the travis build report a number of errors that should be fixed, for example:
https://travis-ci.org/docker/docker-registry/jobs/45954757#L1205

alternatively, you should run the tests suite locally - `python setup.py tests` or `tox` (or `TOX_INCLUDE='' TOX_LIBRARY='' tox`)

estesp commented:
Thanks @dmp42.  I've fixed the tests that were failing, including a potentially important catch for an issue created by the new code--the mocked up images used in a few tests [no idea if it would be hit in real usage, but worth handling properly] had no image data, so attempting to parse that on `GET` and `PUT` repository operations caused a `FileNotFoundException` that was causing an inadvertent code path that was wrong for both cases.  I've fixed that in the code and now all tests are passing.

dmp42 commented:
@shin- @wking what do you think?

wking commented:
On Mon, Jan 05, 2015 at 06:22:14PM -0800, Olivier Gambier wrote:
> @shin- @wking what do you think?

I still want the search result structure in the spec [1].  Having
multiple quasi-official search implementations with different result
structures is just going to be confusing.  Sharing the same endpoint
is not enough for client compatibility ;).

[1]: https://docs.docker.com/reference/api/registry_api/#search

estesp commented:
@wking agreed--might have gotten lost in the noise of earlier comments, but I opened docker/docker#9904 with the change to the search API spec.  Of course I've only succeeded in confusing @jfrazelle so far given these changes aren't available in any registry yet :)

jfrazelle commented:
lol sorry I am caught up now
